### PR TITLE
Feat/actual claw

### PR DESCRIPTION
**Pull request by X-EX3-X**
for justin to check

### CODE REVIEW

File: src/main/cpp/IO/ClawIO.cpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+void ClawIO::SimulationPeriodic() {
+    // SimulationPeriodic
+}
+
+bool ClawIO::ProcessIO() {
+    OKC_CHECK(sw_interface_ != nullptr);
+    OKC_CHECK(hw_interface_ != nullptr);
+
```
Comment (jkleiber) [inline src/main/cpp/IO/ClawIO.cpp:16]: Add `OKC_CHECK(hw_interface_->claw_open_and_close_encoder != nullptr);` here

File: src/main/cpp/IO/ClawIO.cpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    // Set the software outputs
+    // If the encoder should be reset, reset it
+    if (sw_interface_->reset_claw_open_and_close) {
+         hw_interface_->claw_open_and_close_encoder->SetPosition (0);
+        // Lower the encoder reset flag
+        sw_interface_->reset_claw_open_and_close = false;
+    }
+    hw_interface_->claw_open_and_close_motor->Set (sw_interface_->claw_open_and_close_power);
```
Comment (jkleiber) [inline src/main/cpp/IO/ClawIO.cpp:24]: why is there a space between `Set` and `(`

File: src/main/cpp/subsystems/claw.cpp [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#include "subsystems/Claw.h"
+
```
Comment (jkleiber) [inline src/main/cpp/subsystems/claw.cpp:2]: You need to at least put placeholder functions for `Periodic` and `SimulationPeriodic`

File: src/main/include/Utils.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
+
+#pragma once
+
+#include <iostream>
+
```
Comment (jkleiber) [inline src/main/include/Utils.h:5]: This file will conflict with master after #34. Please be careful as to not overwrite that change

File: src/main/include/hardware/Actuators.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+#define RIGHT_MOTOR_1 4
+#define RIGHT_MOTOR_2 5
+#define RIGHT_MOTOR_3 6
+
+#define INDEXER_MOTOR 7
+#define INTAKE_POSITION_MOTOR 10
+#define INTAKE_MOTOR 11
+
+#define SHOOTER_MOTOR 8
+#define TRIGGER_MOTOR 9
```
Comment (jkleiber) [inline src/main/include/hardware/Actuators.h:27]: why are these copied in?

File: src/main/include/hardware/Hardware.h [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+#include <frc/drive/DifferentialDrive.h>
+#include <frc/motorcontrol/MotorControllerGroup.h>
+
+#include "hardware/ActuatorInterface.h"
+#include "hardware/SensorInterface.h"
+
+// Subsystem I/O
+#include "io/DrivetrainIO.h"
+#include "io/IntakeIO.h"
+#include "io/ShooterIO.h"
```
Comment (jkleiber) [inline src/main/include/hardware/Hardware.h:12]: These files don't exist, and should not be included

File: src/main/include/hardware/Hardware.h [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+typedef struct hardware_t {
+    // Actuators
+    std::unique_ptr<ActuatorInterface> actuators;
+
+    // Drivetrain specific hardware abstractions.
+    std::unique_ptr<frc::DifferentialDrive> diff_drive;
```
Comment (jkleiber) [inline src/main/include/hardware/Hardware.h:19]: This shouldn't be included

File: src/main/include/hardware/Hardware.h [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+ * @return false
+ */
+bool SetupClawInterface(
+    std::unique_ptr<HardwareInterface> &hardware,
+    std::shared_ptr<ShooterHardwareInterface> *interface);
+
+ bool SetupVisionTrackingInterface(
+    std::unique_ptr<HardwareInterface> &hardware,
+    std::shared_ptr<ShooterHardwareInterface> *interface);
+
```
Comment (jkleiber) [inline src/main/include/hardware/Hardware.h:63]: Why is this here

File: src/main/include/io/clawIO.h [hunk 9]
```diff
@@ -0,0 +1,52 @@
+#pragma once
```
Comment (jkleiber) [inline src/main/include/io/clawIO.h:1]: This file needs to be named `ClawIO.h` not `clawIO.h`. Capitalize the file names

File: src/main/include/io/clawIO.h [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+
+typedef struct claw_config_t {
+    
+} ClawConfig;
+
+typedef struct claw_hardware_interface_t {
+    rev::CANSparkMax *const claw_open_and_close_motor;
+    rev::SparkMaxRelativeEncoder *const claw_open_and_close_encoder;
+} ClawHardwareInterface;
```
Comment (jkleiber) [inline src/main/include/io/clawIO.h:20]: Let's name this `ClawHardware`

File: src/main/include/subsystems/claw.h [hunk 11]
```diff
@@ -0,0 +1,39 @@
+// Copyright (c) FIRST and other WPILib contributors.
```
Comment (jkleiber) [inline src/main/include/subsystems/claw.h:1]: This file should be `Claw.h` not `claw.h`. Capitalize file names

**Review by jkleiber — CHANGES_REQUESTED** [review]
The bare bones are here, but we need to only include relevant code. Once the swerve branch gets merged, hopefully a lot of these file changes go away - we need to make sure this branch doesn't overwrite hard work done on other branches.

**Review by jkleiber — APPROVED** [review]
Looks good to me @X-EX3-X @danielbrownmsm  feel free to merge

File: src/main/cpp/subsystems/Claw.cpp [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+bool Claw::ResetPositionEncoder(){
+    return true;
+}
+bool Claw::ResetPositionPID(){
+    return true;
+}
+bool Claw::SetPosition(){
+    // SetPosition (30); //cone
+    // SetPosition (0); //close
+    // SetPosition (50);//cube
```
Comment (danielbrownmsm) [inline src/main/cpp/subsystems/Claw.cpp:16]: I mean, really this PR has no functionality, but at least getting the framework is something

**Review by danielbrownmsm — APPROVED** [review]
yeah we can merge. will need to add a lot of stuff, but this is a good skeleton

**Pull request merged**